gateway-js/src/index.ts

```diff
@@ -528,20 +613,49 @@ export class ApolloGateway implements GraphQLService {
     }
   }
 
-  private async updateWithSupergraphSdl(
-    result: SupergraphSdlUpdate,
-  ): Promise<void> {
-    if (result.id === this.compositionId) {
-      this.logger.debug('No change in composition since last check.');
-      return;
+  private externalSupergraphUpdateCallback(supergraphSdl: string) {
+    this.updateWithSupergraphSdl({
+      supergraphSdl,
+      id: this.getIdForSupergraphSdl(supergraphSdl),
+    });
+  }
+
+  private async externalSubgraphHealthCheckCallback(supergraphSdl: string) {
+    const parsedSupergraphSdl =
+      supergraphSdl === this.supergraphSdl
+        ? this.parsedSupergraphSdl
+        : parse(supergraphSdl);
+
+    const serviceList = this.serviceListFromSupergraphSdl(parsedSupergraphSdl!);
+    // Here we need to construct new datasources based on the new schema info
+    // so we can check the health of the services we're _updating to_.
+    const serviceMap = serviceList.reduce((serviceMap, serviceDef) => {
+      serviceMap[serviceDef.name] = {
+        url: serviceDef.url,
+        dataSource: this.createDataSource(serviceDef),
+      };
+      return serviceMap;
+    }, Object.create(null) as DataSourceMap);
+
+    try {
+      await this.serviceHealthCheck(serviceMap);
+    } catch (e) {
+      throw new Error(
+        'The gateway subgraphs health check failed. Updating to the provided ' +
+          '`supergraphSdl` will likely result in future request failures to ' +
+          'subgraphs. The following error occurred during the health check:\n' +
+          e.message,
+      );
     }
+  }
 
+  private updateWithSupergraphSdl({ supergraphSdl, id }: SupergraphSdlUpdate) {
     // TODO(trevor): #580 redundant parse
     // This may throw, so we'll calculate early (specifically before making any updates)
     // In the case that it throws, the gateway will:
     //   * on initial load, throw the error
     //   * on update, log the error and don't update
-    const parsedSupergraphSdl = parse(result.supergraphSdl);
+    const parsedSupergraphSdl = parse(supergraphSdl);
```

Comment:
Bad SDL will lead to the user's update hook throwing. That seems reasonable, right?

Reply:
Reasonable! Will make note of this in the documentation. Also added a logger and error handling in the `IntrospectAndCompose` implementation to account for errors on update.